### PR TITLE
[reflection] cleanup `derive_reflect`

### PR DESCRIPTION
**Pull request by NathanSWard**
# Objective

- The current `derive_reflect` implementation is a large function with difficult to follow code paths.
- When determining the "active fields" (e.g. those that don't use `#[reflect(ignore)]`) the logic is unnecessarily complex.

## Solution

- Refactor `derive_reflect` by breaking into into smaller functions, and simplifying the logic for obtaining the "active fields".

### CODE REVIEW

**Review by bjorn3 — APPROVED** [review]

File: crates/bevy_reflect/bevy_reflect_derive/src/lib.rs [hunk 1]
```diff
@@ -65,66 +96,14 @@ pub fn derive_reflect(input: TokenStream) -> TokenStream {
...
-        .filter(|(_field, attrs, _i)| {
-            attrs.is_none()
-                || match attrs.as_ref().unwrap().ignore {
-                    Some(ignore) => !ignore,
-                    None => true,
-                }
-        })
-        .map(|(f, _attr, i)| (*f, *i))
-        .collect::<Vec<(&Field, usize)>>();
+    let active_fields = active_fields(&fields).collect::<Vec<_>>();
```
Comment (Davier) [inline crates/bevy_reflect/bevy_reflect_derive/src/lib.rs:99]: This merges two steps into one which is good, but I will probably need to undo it in #1395 in order to also get the list of inactive fields (see https://github.com/bevyengine/bevy/pull/1395/files#diff-786606f1541ef395577b541509d2834381796a7f9859e6aaeef68dcf551928bcR211-R220 ). That's my problem I guess :-)

**Comment by alice-i-cecile** [comment]
@MrGVSV @jakobhellermann can I get a review on this from one of you? I think this is an easy win.

**Review by alice-i-cecile — APPROVED** [review]

**Review by MrGVSV — APPROVED** [review]
Nice and simple. Personally, I'd like more internal documentation on the added functions to help out future contributors, but it's not a blocking issue.

**Comment by alice-i-cecile** [comment]
Because Nathan has been busy lately, I'm going to bias towards doing follow-up in future PRs :)

**Comment by alice-i-cecile** [comment]
@NathanSWard, any chance you'll have a chance to fix this up in the next week or so? If not, no worries. We'll open another PR off this and credit you.

**Comment by PROMETHIA-27** [comment]
Is this PR made obsolete by #4712?

**Comment by alice-i-cecile** [comment]
Agreed. I'm going to close this out in favor of #4712. @MrGVSV can you credit the work done here in your PR description :)

Nathan did great work here; it's our bad for having let it sit in backlog for so long.

**Comment by MrGVSV** [comment]
> @MrGVSV please credit the work done here in your PR description in whatever form you see fit :)

Yeah that's my bad for not recognizing this prior work already. I'll definitely add a note to the PR description tonight. Thanks for pointing this out!

**Pull request closed without merging**